flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// build v2
+		stateMap.put(1, 1L, "2");
+		referenceValues.add(0, "2");
+
+		// take snapshot3
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot3 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// build v3
+		stateMap.put(1, 1L, "3");
+		referenceValues.add(0, "3");
+
+		// take snapshot4
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot4 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// take snapshot5
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot5 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		// take snapshot6
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot6 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+
+		assertEquals(6, stateMap.getStateMapVersion());
+		assertEquals(6, stateMap.getHighestRequiredSnapshotVersion());
+		assertEquals(6, stateMap.getSnapshotVersions().size());
+		assertEquals(5, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+
+		Map<Long, Map<Integer, String>> referenceStates = new HashMap<>();
+		referenceStates.put(1L, new HashMap<>());
+		referenceStates.get(1L).put(1, "0");
+
+		// complete snapshot 1, and no value will be removed
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot1, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot1.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(5, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(1, stateMap.getHighestFinishedSnapshotVersion());
+
+		// complete snapshot 3, and v0 will be removed
+		referenceStates.get(1L).put(1, "2");
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot3, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot3.release();
+		referenceValues.remove(referenceValues.size() - 1);
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(1, stateMap.getHighestFinishedSnapshotVersion());
+
+		// complete snapshot 2, and no value will be removed
+		referenceStates.get(1L).put(1, "1");
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot2, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot2.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(3, stateMap.getHighestFinishedSnapshotVersion());
+
+		// add node to pruning set to prevent snapshot4 to prune
+		stateMap.getPruningValueNodes().add(node);
```

Comment:
I don't like this call because it accesses internal state of the `stateMap` which breaks encapsulation. Is this really needed?

Reply:
Yes, pruning out-of-date versions of value w/o any snapshot reference is an internal processing and we'd like to assure the correctness of the logic here in this test case.

Reply:
I think this will bite us later but if you say there is no other way, I guess we need to take the bullet.